### ISSUE TITLE
Efficient computation of intersection and volume.

### ISSUE THREAD

**Issue by dhruvdcoder** [iesl/box-embeddings#32]
Once we have minus_Z and a parameterization that stores minuz_Z instead of Z directly, we can use z and minus_Z to compute intersection and volume.

**Comment by mboratko** [comment]
```python
class PureGumbelBox(Module):
    """
    Box embedding model optimized for performance.

    In this model, the self.boxes parameter is of shape (num_entity, 2, dim), where self.boxes[i,:,k] are location
    parameters for Gumbel distributions representing the corners of the ith box in the kth dimension.
        self.boxes[i,0,k] is the location parameter mu_z for a MaxGumbel distribution
        self.boxes[i,1,k] represents -mu_Z, i.e. negation of location parameter, for a MinGumbel distribution
    This rather odd convention is chosen to maximize performance.

    Note that with this parameterization, we allow the location parameter to "flip around", i.e. mu_z > mu_Z.
    This is completely reasonable, from the GumbelBox perspective.
    """

    def __init__(
        self,
        num_entity,
        dim,
        volume_temp=1.0,
        intersection_temp=1.0,
    ):
        super().__init__()
        self.boxes = Parameter(
            torch.sort(
                torch.randn((num_entity, 2, dim)), dim=-2, descending=True
            ).values
        )

        self.volume_temp = volume_temp
        self.intersection_temp = intersection_temp
        self.softplus = torch.nn.Softplus(beta=1 / self.volume_temp)
        self.softplus_const = 2 * self.intersection_temp * 0.57721566490153286060

    def forward(self, idxs: LongTensor) -> Tensor:
        """
        A version of the forward pass that is slightly more performant.
        :param idxs: Tensor of shape (..., 2) indicating edges, i.e. [...,0] -> [..., 1] is an edge
        :return:
        """
        boxes = self.boxes[idxs]
        # we now have boxes in a shape (..., nodeA/nodeB, min/max, embedding_dim)

        # calculate Gumbel intersection
        intersection = self.intersection_temp * torch.logsumexp(
            boxes / self.intersection_temp, dim=-3, keepdim=True
        )
        intersection = torch.max(
            torch.cat((intersection, boxes), dim=-3), dim=-3
        ).values
        # calculating log volumes
        # keep in mind that the [...,1,:] represents negative max, thus we negate it
        log_intersection_volume = torch.sum(
            torch.log(
                self.softplus(-intersection.sum(dim=-2) - self.softplus_const) + 1e-23
            ),
            dim=-1,
        )

        return log_intersection_volume
```

**Comment by mboratko** [comment]
1. Setup two scripts, one which uses the above "efficient" computation and one which uses a similar computation but treats the parameters as `mu_z` and `mu_Z`, which will require negating and stacking the `mu_Z` parameter (like we currently do).
2. Profile these scripts (eg. using timeit)
3. If treating the parameters as though they are `mu_z` and `-mu_Z` is faster, we should consider making this the new base class.

**Comment by mboratko** [comment]
```
class PureGumbelBoxNormal(Module):
    """
    Box embedding model.

    In this model, the self.boxes parameter is of shape (num_entity, 2, dim), where self.boxes[i,:,k] are location
    parameters for Gumbel distributions representing the corners of the ith box in the kth dimension.
        self.boxes[i,0,k] is the location parameter mu_z for a MaxGumbel distribution
        self.boxes[i,1,k] is the location parameter mu_z for a MaxGumbel distribution

    Note that with this parameterization, we allow the location parameter to "flip around", i.e. mu_z > mu_Z.
    This is completely reasonable, from the GumbelBox perspective.
    """

    def __init__(
        self,
        num_entity,
        dim,
        volume_temp=1.0,
        intersection_temp=1.0,
    ):
        super().__init__()
        self.boxes = Parameter(
            torch.sort(
                torch.randn((num_entity, 2, dim)), dim=-2, descending=True
            ).values
        )

        self.volume_temp = volume_temp
        self.intersection_temp = intersection_temp
        self.softplus = torch.nn.Softplus(beta=1 / self.volume_temp)
        self.softplus_const = 2 * self.intersection_temp * 0.57721566490153286060

    def forward(self, idxs: LongTensor) -> Tensor:
        """
        A version of the forward pass that is slightly more performant.
        :param idxs: Tensor of shape (..., 2) indicating edges, i.e. [...,0] -> [..., 1] is an edge
        :return:
        """
        boxes = self.boxes[idxs]
        # we now have boxes in a shape (..., nodeA/nodeB, min/max, embedding_dim)

        # calculate Gumbel intersection
        intersection_min = self.intersection_temp * torch.logsumexp(
            boxes[...,0,:] / self.intersection_temp, dim=-3, keepdim=True
        )
        intersection_min = torch.max(intersection_min, boxes[...,0,:])
        intersection_max = -self.intersection_temp * torch.logsumexp(
            -boxes[...,1,:] / self.intersection_temp, dim=-3, keepdim=True
        )
        intersection_max = torch.min(intersection_max, boxes[...,1,:])
        # calculating log volumes
        # keep in mind that the [...,1,:] represents negative max, thus we negate it
        log_intersection_volume = torch.sum(
            torch.log(
                self.softplus(intersection_max-intersection_min - self.softplus_const) + 1e-23
            ),
            dim=-1,
        )

        return log_intersection_volume
```

**Comment by mboratko** [comment]
Suggested changes:
1. in the `_compute` functions, stack `z` and `-Z`, run a single `logaddexp`, and then (for the clipping functions) do `torch.max`, finally return `lse[...,0,:]` and `-lse[...,1,:]`
2. `gumbel _intersection` should be returning a base box
3. (optional) for the clipping with different forward / backward pass, we should probably implement this by subclassing `Function` as opposed to using the `.detach` trick.

**Comment by purujitgoyal** [comment]
PureGumbleBox (Stacking both z and Z in one tensor)
	CPU:
		54 s ± 467 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

		Line #    Mem usage    Increment  Occurences   Line Contents
		============================================================
		   123    125.8 MiB    125.8 MiB           1   @profile    
		   124                                         def train(gumble, optimizer: optim.SGD, idx, n_epoch=10000):
		   125    140.1 MiB      0.0 MiB       10001       for epoch in range(n_epoch):
		   126    140.1 MiB      0.0 MiB       10000           optimizer.zero_grad()
		   127    140.1 MiB      6.0 MiB       10000           loss = -gumble(idx)
		   128    140.1 MiB      4.4 MiB       10000           loss.backward()
		   129    140.1 MiB      3.9 MiB       10000           optimizer.step()


	GPU:
		23.8 s ± 1.14 s per loop (mean ± std. dev. of 7 runs, 1 loop each)

		Line #    Mem usage    Increment  Occurences   Line Contents
		============================================================
		   123   2002.2 MiB   2002.2 MiB           1   @profile    
		   124                                         def train(gumble: PureGumbelBox, optimizer: optim.SGD, idx, n_epoch=10000):
		   125   2002.2 MiB      0.0 MiB       10001       for epoch in range(n_epoch):
		   126   2002.2 MiB      0.0 MiB       10000           optimizer.zero_grad()
		   127   2002.2 MiB      0.0 MiB       10000           loss = -gumble(idx)
		   128   2002.2 MiB      0.0 MiB       10000           loss.backward()
		   129   2002.2 MiB      0.0 MiB       10000           optimizer.step()




PureGumbelBoxNormal (not stacking):
	
	CPU:
		52.6 s ± 521 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

		Line #    Mem usage    Increment  Occurences   Line Contents
		============================================================
		   123    118.3 MiB    118.3 MiB           1   @profile    
		   124                                         def train(gumble, optimizer: optim.SGD, idx, n_epoch=10000):
		   125    137.6 MiB      0.0 MiB       10001       for epoch in range(n_epoch):
		   126    137.6 MiB      0.0 MiB       10000           optimizer.zero_grad()
		   127    137.6 MiB      3.3 MiB       10000           loss = -gumble(idx)
		   128    137.6 MiB      8.2 MiB       10000           loss.backward()
		   129    137.6 MiB      7.7 MiB       10000           optimizer.step()


	GPU:
 		28.1 s ± 1.38 s per loop (mean ± std. dev. of 7 runs, 1 loop each)

 		Line #    Mem usage    Increment  Occurences   Line Contents
		============================================================
		   123   2002.2 MiB   2002.2 MiB           1   @profile    
		   124                                         def train(gumble: PureGumbelBox, optimizer: optim.SGD, idx, n_epoch=10000):
		   125   2002.2 MiB      0.0 MiB       10001       for epoch in range(n_epoch):
		   126   2002.2 MiB      0.0 MiB       10000           optimizer.zero_grad()
		   127   2002.2 MiB      0.0 MiB       10000           loss = -gumble(idx)
		   128   2002.2 MiB      0.0 MiB       10000           loss.backward()
		   129   2002.2 MiB      0.0 MiB       10000           optimizer.step()
